aspnetcore/signalr/authn-and-authz.md

```diff
@@ -0,0 +1,114 @@
+
+### Bearer Token Authentication
+
+When using the .NET Client, or when your SignalR Hubs are located on a different server from your web application, we recommend using bearer token authentication. In this authentication scheme, you configure the SignalR client with an access token to send to the server. The server validates this token and uses the data within it to identify the user. The details of bearer token authentication are beyond the scope of this document, but there are some things to note when using this form of authentication in SignalR. On the server, bearer token authentication is configured using the [JWT Bearer middleware](/dotnet/api/microsoft.extensions.dependencyinjection.jwtbearerextensions.addjwtbearer?view=aspnetcore-2.1).
+
+In the JavaScript client, the token can be provided using the `accessTokenFactory` option.
+
+[!code-javascript[Configure Access Token](authn-and-authz/sample/wwwroot/js/chat.ts?range=63-65)]
+
+In the .NET client, there is a simlar `AccessTokenProvider` property that can be used to configure this:
+
+```csharp
+var connection = new HubConnectionBuilder()
+    .WithUrl("https://example.com/myhub", options => options.AccessTokenProvider = () => _myAccessToken)
+    .Build();
+```
+
+> [!NOTE]
+> This function is called before **every** HTTP request made by SignalR. This means that if you need to renew the token in order to keep the connection active (because it may expire during the connection) you can do so from within this function and SignalR will automatically start using the new token.
+
+In standard Web APIs, bearer tokens are sent via an HTTP header. However, due to security limitations imposed by browsers, the browser client is not able to set these headers when using the WebSockets or Server-Sent Events transports. When using those transports, the token is transmitted as a query string parameter. In order to support this on the server, additional configuration is required:
+
+[!code-csharp[Configure Server to accept access token from Query String](authn-and-authz/sample/Startup.cs?range=33-34,42-80,90)]
+
+### Windows Authentication
+
+If you have configured [Windows Authentication](xref:security/authentication/windowsauth) in your application, SignalR can use that authentication to secure Hubs. However, in order to send messages to individual users, you need to add a custom User ID provider, because the Windows Authentication system does not provide the "Name Identifier" claim that SignalR uses to determine the user name. To do this, add a new class that implements `IUserIdProvider` and retrieve one of the claims from the user to use as the identifier. For example, to use the "Name" claim (which is the Windows username in the form `[Domain]\[Username]`), create the following class:
```

Comment:
application->app
Hubs->hubs
However, in order to send messages to individual users, you need to add a custom User ID provider, because the Windows Authentication system does not provide the "Name Identifier" claim that SignalR uses to determine the user name.
->
However, In order to send messages to individual users, you need to add a custom user ID provider. This is because the Windows authentication system doesn't provide the "Name Identifier" claim that SignalR uses to determine the user name.